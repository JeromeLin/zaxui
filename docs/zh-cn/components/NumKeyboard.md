# 数字键盘 NumKeyboard

[demo页面](https://zhongantecheng.github.io/zarm/#/numkeyboard)

### 引入

```js
import { NumKeyboard } from 'zarm';
```


### 代码演示

#### 基本用法

##### 手机号码键盘
```jsx
<NumKeyboard type="tel"/>
```

##### 价格键盘
```jsx
<NumKeyboard type="price"/>
```

##### 身份证键盘
```jsx
<NumKeyboard type="id"/>
```

### API

| 属性 | 类型 | 默认值 | 可选值／参数 | 说明 |
| :--- | :--- | :--- | :--- | :--- |
| prefixCls | string | za-input | | 类名前缀 |
| className | string | | | 追加类名 |
| type | string | tel | | 键盘类型 |
| mask | bool | false | | 是否有蒙版 |
| btnTitle | string | 确定 | | 确定文案 |
| visible | bool | false | | 键盘是否可见 |
| keyCallback | func | | | 键盘数字按钮的回调 |
| doneCallback | func |  | | 键盘确定按钮的回调 |