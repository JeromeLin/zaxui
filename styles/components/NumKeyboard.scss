$prefixCls: $prefixCls-numkeyboard;
$grey: #f5f5f5;
$white: #fff;
$blue: #249fe6;
$lineColor: #e6e6e6;
$keyColor: #333;

.#{$prefixCls} {
  // position: fixed;
  // z-index: 9999;
  width: 100%;
  transition: all .2s ease-out;
  .op-zone {
    background-color: $grey;
    font-size: r(18);
    padding: r(10) 0;
    overflow: hidden;
    &:before {
      content: '';
      display: block;
      width: 100%;
      border-top: 1px solid $lineColor;
      transform: scaleY(.5);
      position: absolute;
      top: 0;
    }
    .comfirm-btn {
      margin-right: r(15);
      color: $blue;
      float: right;
      background: transparent;
      border: 0;
      outline: none;
    }
  }
  .key-zone {
    height: r(220);
    background: $white;
    font-size: r(24);
    color: $keyColor;
    display: flex;
    flex-wrap: wrap;
    .key-btn {
      position: relative;
      width: 33.33%;
      text-align: center;
      line-height: r(55);
      background-color: $white;
      &:nth-of-type(10) {
        background-color: $grey;
      }
      &:nth-of-type(12) {
        background-color: $grey;
      }
      &:active{
        box-shadow: inset 0 r(4) 0 $grey,
                    inset 0 r(-4) 0 $grey,
                    inset r(4) 0 0 $grey,
                    inset r(-4) 0 0 $grey;
      }
      &.top-1px-line {
        @include onepx(top, $lineColor);
      }
      &.top-right-1px-line {
        @include onepx(top right, $lineColor);
      }
      // &.top-1px-line {
        // &:before {
        //   content: '';
        //   display: block;
        //   width: 100%;
        //   height: 0;
        //   position: absolute;
        //   top: 0;
        //   border-top: 1px solid $lineColor;
        //   transform: scaleY(.5);
        // }
      // }
      // &.right-1px-line {
      //   @include onepx(right, $lineColor);
        // &:after {
        //   content: '';
        //   display: block;
        //   width: 0;
        //   height: 100%;
        //   position: absolute;
        //   top: 0;
        //   right: 0;
        //   border-right: 1px solid $lineColor;
        //   transform: scaleX(.5);
        // }
      // }
      // .backspace-logo {
      //   display: inline-block;
      //   width: r(25);
      //   height: r(18);
      //   line-height: r(18);
      //   margin: auto;
      //   background-image: url(./backspace.png);
      //   background-size: 100% 100%;
      // }
    }
  }

}