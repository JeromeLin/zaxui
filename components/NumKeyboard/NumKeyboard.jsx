import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import FastClick from 'fastclick';
import { Popup, Icon } from '../../components';

const TEL_KEYS = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '', '0', '-1'];
const PRICE_KEYS = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '.', '0', '-1'];
const ID_KEYS = ['1', '2', '3', '4', '5', '6', '7', '8', '9', 'X', '0', '-1'];

class NumKeyboard extends Component {
  componentDidMount() {
    FastClick.attach(this.keyboard, { tapDelay: 10 });
  }
  _genKey() {
    const { type, keyCallback } = this.props;
    let keys = null;
    switch (type) {
      case 'price':
        keys = [].concat(PRICE_KEYS);
        break;
      case 'id':
        keys = [].concat(ID_KEYS);
        break;
      case 'tel':
      default:
        keys = [].concat(TEL_KEYS);
        break;
    }
    return keys.map((key, index) => {
      return (
        <div
          className={classnames('key-btn', { 'top-right-1px-line': index % 3 !== 2, 'top-1px-line': index % 3 === 2 })}
          onClick={() => typeof keyCallback === 'function' && keyCallback(key)}
          key={key}>
          {key !== '-1' ? key : (<Icon type="wrong" />)}
        </div>
      );
    });
  }
  render() {
    const { prefixCls, className, type, mask, btnTitle, visible, doneCallback, keyCallback, ...others } = this.props;
    return (
      <Popup mask={mask} visible={visible} onMaskClick={() => typeof doneCallback === 'function' && doneCallback()}>
        <div className={classnames(`${prefixCls}`, className)} ref={(e) => { this.keyboard = e; }} {...others}>
          <div className="op-zone">
            <button
              className="comfirm-btn"
              onClick={
                () => {
                  typeof doneCallback === 'function' && doneCallback();
                }
              }>{btnTitle}</button>
          </div>
          <div className="key-zone">
            {this._genKey()}
          </div>
        </div>
      </Popup>
    );
  }
}

NumKeyboard.propTypes = {
  prefixCls: PropTypes.string,
  type: PropTypes.string,
  mask: PropTypes.bool,
  btnTitle: PropTypes.string,
  visible: PropTypes.bool,
  keyCallback: PropTypes.func,
  doneCallback: PropTypes.func,
};

NumKeyboard.defaultProps = {
  prefixCls: 'za-numkeyboard',
  mask: false,
  type: 'tel',
  btnTitle: '确定',
  visible: false,
};

export default NumKeyboard;
