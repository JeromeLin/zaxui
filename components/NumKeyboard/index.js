import NumKeyboard from './NumKeyboard';
export default NumKeyboard;