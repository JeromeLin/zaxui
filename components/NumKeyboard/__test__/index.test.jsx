import React from 'react';
import { render, shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import NumKeyboard from '../index';

describe('NumKeyboard', () => {
  const props = {
    type: 'tle',
    visible: true,
  };

  it('render correctly', () => {
    const wrapper = render(<NumKeyboard {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('key button press', () => {
    const keyCallback = jest.fn();
    const wrapper = shallow(<NumKeyboard keyCallback={keyCallback} {...props} />);
    (wrapper.find('.key-btn').first()).simulate('click');
    expect(keyCallback).toBeCalled();
  });

  it('done button press', () => {
    const doneCallback = jest.fn();
    const wrapper = shallow(<NumKeyboard doneCallback={doneCallback} {...props} />);
    (wrapper.find('.op-zone').childAt(0)).simulate('click');
    expect(doneCallback).toBeCalled();
  });
});