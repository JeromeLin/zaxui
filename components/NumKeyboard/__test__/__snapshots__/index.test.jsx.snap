// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumKeyboard render correctly 1`] = `
<div
  class="za-popup"
>
  <div
    class="za-popup-wrapper za-popup-wrapper-bottom"
    style="-webkit-transition-duration:200ms;transition-duration:200ms;"
  >
    <div
      class="za-numkeyboard"
    >
      <div
        class="op-zone"
      >
        <button
          class="comfirm-btn"
        >
          确定
        </button>
      </div>
      <div
        class="key-zone"
      >
        <div
          class="key-btn top-right-1px-line"
        >
          1
        </div>
        <div
          class="key-btn top-right-1px-line"
        >
          2
        </div>
        <div
          class="key-btn top-1px-line"
        >
          3
        </div>
        <div
          class="key-btn top-right-1px-line"
        >
          4
        </div>
        <div
          class="key-btn top-right-1px-line"
        >
          5
        </div>
        <div
          class="key-btn top-1px-line"
        >
          6
        </div>
        <div
          class="key-btn top-right-1px-line"
        >
          7
        </div>
        <div
          class="key-btn top-right-1px-line"
        >
          8
        </div>
        <div
          class="key-btn top-1px-line"
        >
          9
        </div>
        <div
          class="key-btn top-right-1px-line"
        />
        <div
          class="key-btn top-right-1px-line"
        >
          0
        </div>
        <div
          class="key-btn top-1px-line"
        >
          <i
            class="za-icon za-icon-wrong"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
